### PR TITLE
Fix in runonce and jacobi drivers so that they collectively reraise analysis errors detected on a subsystem in a parallel group.

### PR DESCRIPTION
**Pull request by Kenneth-T-Moore**

### CODE REVIEW

**Comment by Kenneth-T-Moore** [comment]
Closed until we discuss this on Wednesday.

File: openmdao/solvers/nonlinear/nonlinear_block_jac.py [hunk 1]
```diff
@@ -41,3 +51,16 @@ def _mpi_print_header(self):
                 header += prefix + pathname + "\n"
                 header += prefix + nchar * "="
                 print(header)
+
+    def _run_apply(self):
+        """
+        Run the the apply_nonlinear method on the system.
```
[the remark this answers is not in the document]
Comment (Kenneth-T-Moore) [reply openmdao/solvers/nonlinear/nonlinear_block_jac.py]: Did a global search and fixed every "the the" in the docs and comments.

File: openmdao/solvers/nonlinear/nonlinear_block_jac.py [hunk 2]
```diff
@@ -41,3 +51,16 @@ def _mpi_print_header(self):
...
+
+    def _run_apply(self):
+        """
+        Run the apply_nonlinear method on the system.
+        """
+        system = self._system
+
+        # If this is a parallel group, check for analysis errors and reraise.
+        if len(system._subsystems_myproc) != len(system._subsystems_allprocs):
+            with MultiProcFailCheck(self._system.comm):
```
Comment (naylor-b) [inline openmdao/solvers/nonlinear/nonlinear_block_jac.py]: forgot to use the 'system' temporary variable here

File: openmdao/solvers/nonlinear/nonlinear_runonce.py [hunk 3]
```diff
@@ -36,9 +37,13 @@ def solve(self):
             # If this is a parallel group, transfer all at once then run each subsystem.
             if len(system._subsystems_myproc) != len(system._subsystems_allprocs):
                 system._transfer('nonlinear', 'fwd')
-                for subsys in system._subsystems_myproc:
-                    subsys._solve_nonlinear()
+
+                with MultiProcFailCheck(self._system.comm):
```
Comment (naylor-b) [inline openmdao/solvers/nonlinear/nonlinear_runonce.py]: replace self._system with system

File: openmdao/utils/mpi.py [hunk 4]
```diff
@@ -123,6 +127,38 @@ def __init__(self):
...
+    Raise an AnalysisError on all procs if it is raised on one.
+
+    Wrap this around code that you want to globally fail if it fails
+    on any MPI process in comm.  If not running under MPI, don't
+    handle any exceptions.
+
+    Parameters
+    ----------
+    comm : MPI communicator or None
+        Communicator from the ParallelSystem that owns the calling solver.
```
Comment (naylor-b) [inline openmdao/utils/mpi.py]: ParallelSystem?

File: openmdao/solvers/nonlinear/tests/test_nonlinear_runonce.py [hunk 5]
```diff
@@ -54,5 +62,72 @@ def test_feature_solver(self):
...
+
+@unittest.skipUnless(PETScVector, "PETSc is required.")
+class TestNonlinearRunOnceSolverMPI(unittest.TestCase):
+
+    N_PROCS = 2
+
+    @unittest.skipUnless(MPI, "MPI is not active.")
+    def test_reraise_analylsis_error(self):
+
+        class AEComp(ExplicitComponent):
```
Comment (JustinSGray) [inline openmdao/solvers/nonlinear/tests/test_nonlinear_runonce.py]: nitpick here, but you defined the same component and driver classes in in the above test. Why not move the class definitions out of the test methods?

File: openmdao/solvers/nonlinear/nonlinear_block_jac.py (hunk 2 above)
Comment (Kenneth-T-Moore) [reply openmdao/solvers/nonlinear/nonlinear_block_jac.py]: Fixed

File: openmdao/solvers/nonlinear/nonlinear_runonce.py (hunk 3 above)
Comment (Kenneth-T-Moore) [reply openmdao/solvers/nonlinear/nonlinear_runonce.py]: Fixed

File: openmdao/utils/mpi.py (hunk 4 above)
Comment (Kenneth-T-Moore) [reply openmdao/utils/mpi.py]: Fixed

File: openmdao/solvers/nonlinear/tests/test_nonlinear_runonce.py (hunk 5 above)
Comment (Kenneth-T-Moore) [reply openmdao/solvers/nonlinear/tests/test_nonlinear_runonce.py]: Done

File: openmdao/utils/mpi.py [hunk 6]
```diff
@@ -123,6 +127,38 @@ def __init__(self):
         self.size = 1
 
 
+@contextmanager
+def MultiProcFailCheck(comm):
```
Comment (hwangjt) [inline openmdao/utils/mpi.py]: minor detail, but should this be snake case since it is a function?

Comment (Kenneth-T-Moore) [reply openmdao/utils/mpi.py]: Fixed

**Pull request merged**